benchmarks/src/priority_queue_push_range.cpp

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <benchmark/benchmark.h>
+#include <functional>
+#include <queue>
+#include <random>
+#include <span>
+#include <string>
+#include <vector>
+using namespace std;
+
+namespace {
+    constexpr size_t vec_size = 8000;
+
+    template <class T>
+    auto create_vec(size_t vsize, function<T(uint64_t)> transform) {
+        vector<T> vec(vsize);
+        for (mt19937_64 rnd(1); auto& e : vec) {
+            e = transform(rnd());
+        }
+        return vec;
+    }
+
+    template <class T>
+    T cast_to(uint64_t val) {
+        return static_cast<T>(val);
+    }
+
+    const auto vec_u8     = create_vec<uint8_t>(vec_size, cast_to<uint8_t>);
+    const auto vec_u16    = create_vec<uint16_t>(vec_size, cast_to<uint16_t>);
+    const auto vec_u32    = create_vec<uint32_t>(vec_size, cast_to<uint32_t>);
+    const auto vec_u64    = create_vec<uint64_t>(vec_size, cast_to<uint64_t>);
+    const auto vec_float  = create_vec<float>(vec_size, cast_to<float>);
+    const auto vec_double = create_vec<double>(vec_size, cast_to<double>);
+
+    const auto vec_str =
+        create_vec<string>(vec_size, [](uint64_t val) { return to_string(static_cast<uint32_t>(val)); });
+    const auto vec_wstr =
+        create_vec<wstring>(vec_size, [](uint64_t val) { return to_wstring(static_cast<uint32_t>(val)); });
+
+    template <class T, const auto& Data>
+    void BM_push_range(benchmark::State& state) {
+        const size_t frag_size = static_cast<size_t>(state.range(0));
+
+        for (auto _ : state) {
+            priority_queue<T> que;
+            span spn{Data};
+
+            while (!spn.empty()) {
+                size_t take_size = min(spn.size(), frag_size);
+                que.push_range(spn.subspan(0, take_size));
+                spn = spn.subspan(take_size);
+            }
+            benchmark::DoNotOptimize(que);
+        }
+    }
+
+    template <size_t L>
+    void putln(const benchmark::State&) {
+        static bool b = [] {
+            puts("");
+            return true;
+        }();
+    }
+} // namespace
+
+#define TEST_PUSH_RANGE(T, source)      \
+    BENCHMARK(BM_push_range<T, source>) \
+        ->Setup(putln<__LINE__>)        \
+        ->RangeMultiplier(100)          \
+        ->Range(1, vec_size)            \
+        ->Arg(vec_size / 2 + 1);

```

Comment:
`vec_size/2+1` will make the function firstly `make_heap` the first half of data, then `push_heap` the rest half of data.